packages/mdc-ripple/_mixins.scss

```diff
@@ -53,64 +55,120 @@ $mdc-ripple-common-styles-emitted_: false !default;
   // without needing to re-include mdc-ripple-surface.
   &::before,
   &::after {
-    @include mdc-ripple-color_($color, $opacity);
+    @include mdc-theme-prop(background-color, $color, $edgeOptOut: true);
 
     opacity: 0;
   }
+}
 
-  // Note: when :active is applied, :focus is already applied, which will effectively double the effect.
-  &:not(.mdc-ripple-upgraded) {
-    &:hover::before,
-    &:focus::before,
-    &:active::after {
-      transition-duration: 85ms;
-      opacity: .6;
-    }
+@mixin mdc-states-hover-opacity($opacity) {
+  // Background wash styles, for both CSS-only and upgraded stateful surfaces
+  &:hover::before {
+    opacity: $opacity;
   }
+}
+
+@mixin mdc-states-focus-opacity($opacity, $include-focus-within: false) {
+  // Focus overrides hover by reusing the ::before pseudo-element.
+  // :focus-within generally works on non-MS browsers and matches when a *child* of the element has focus.
+  // It is useful for cases where a component has a focusable element within the root node, e.g. text field,
+  // but undesirable in general in case of nested stateful components.
+  // We use a modifier class for JS-enabled surfaces to support all use cases in all browsers.
+  $cssOnlyFocusSelector: if(
+    $include-focus-within,
+    "&:not(.mdc-ripple-upgraded):focus::before, &:not(.mdc-ripple-upgraded):focus-within::before",
+    "&:not(.mdc-ripple-upgraded):focus::before"
+  );
 
+  #{$cssOnlyFocusSelector},
   &.mdc-ripple-upgraded--background-focused::before {
-    opacity: .99999;
+    // Note that this duration is only effective on focus, not blur
+    transition-duration: 75ms;
+    opacity: $opacity;
   }
+}
 
-  &.mdc-ripple-upgraded--background-active-fill::before {
-    transition-duration: 120ms;
-    opacity: 1;
-  }
+@mixin mdc-states-press-opacity($opacity) {
+  // Styles for non-upgraded (CSS-only) stateful surfaces
 
-  // Foreground ripple styles
+  &:not(.mdc-ripple-upgraded) {
+    // Apply press additively by using the ::after pseudo-element
+    &::after {
+      transition: opacity $mdc-ripple-fade-out-duration linear;
+    }
 
-  &.mdc-ripple-upgraded::after {
-    opacity: 0;
+    &:active::after {
+      transition-duration: $mdc-ripple-fade-in-duration;
+      opacity: $opacity;
+    }
   }
 
-  &.mdc-ripple-upgraded--foreground-activation::after {
-    animation: 300ms mdc-ripple-fg-radius-in forwards, 83ms mdc-ripple-fg-opacity-in forwards;
+  &.mdc-ripple-upgraded {
+    --mdc-ripple-fg-opacity: $opacity;
   }
+}
 
-  &.mdc-ripple-upgraded--foreground-deactivation::after {
-    animation: 83ms mdc-ripple-fg-opacity-out;
-    // Retain transform from mdc-ripple-fg-radius-in activation
-    transform: translate(var(--mdc-ripple-fg-translate-end, 0)) scale(var(--mdc-ripple-fg-scale, 1));
-  }
+// Simple mixin which automatically selects opacity values based on whether the ink color is light or dark.
+@mixin mdc-states-color($color: black, $include-focus-within: false) {
```

Comment:
hmm okay I think I see the difference between base-color and color now....

What if we renamed mdc-states-base-color to just mdc-states-color. And then renamed mdc-states-color to mdc-general-states. I talked to Michelle and she can add the term "general states" to the article to cover hover, focused, and pressed

Reply:
Before we decide on that, I should probably point out that I was planning to also include optional flags in this mixin for optionally emitting styles for activated and selected states. Would that term still apply then?

Reply:
We talked offline some. I thing mdc-state-hover-opacity is clear...and mdc-states for the "simple" mixins is also clear.